pipenv/patched/pew/pew.py

```diff
@@ -148,79 +148,123 @@ def inve(env, command, *args, **kwargs):
+            # subshell in an error. This causes the subprocess to fail, which
+            # is not what we want. Stay silent for them.
+            name, _ = os.path.splitext(os.path.basename(self.cmd))
+            suppress_error = name.lower() in ('cmd', 'powershell', 'pwsh')
+            if not suppress_error:
+                raise
+
+
+class Bash(Shell):
+
+    # Bash is a special little snowflake, and prevent_path_errors cannot work.
     # https://github.com/berdario/pew/issues/58#issuecomment-102182346
-    bashrcpath = expandpath('~/.bashrc')
-    if bashrcpath.exists():
-        with NamedTemporaryFile('w+') as rcfile:
-            with bashrcpath.open() as bashrc:
-                rcfile.write(bashrc.read())
-            rcfile.write('\nexport PATH="' + to_unicode(compute_path(env)) + '"')
-            rcfile.flush()
-            fork_shell(env, ['bash', '--rcfile', rcfile.name], cwd)
-    else:
-        fork_shell(env, ['bash'], cwd)
+    path_errors_prevented = False
+
+    def fork(self, env, cwd):
+        bashrcpath = expandpath('~/.bashrc')
+        if bashrcpath.exists():
+            with NamedTemporaryFile('w+') as rcfile:
+                with bashrcpath.open() as bashrc:
+                    rcfile.write(bashrc.read())
+                rcfile.write('\nexport PATH="{0}"'.format(
+                    to_unicode(compute_path(env)),
+                ))
+                rcfile.flush()
+                self.args.extend(['--rcfile', rcfile.name])
+                super(Bash, self).fork(env, cwd)
+        else:
+            super(Bash, self).fork(env, cwd)
+
 
+class MicrosoftShell(Shell):
+    # On Windows the PATH is usually set with System Utility
+    # so we won't worry about trying to check mistakes there.
+    path_errors_prevented = False
+
+
+class CmderEmulatedShell(MicrosoftShell):
+    def fork(self, env, cwd):
+        if cwd:
+            os.environ['CMDER_START'] = cwd
+        super(CmderEmulatedShell, self).fork(env, cwd)
+
+
+class CmderCommandPrompt(CmderEmulatedShell):
+    def fork(self, env, cwd):
+        rc = '%CMDER_ROOT%\\vendor\\init.bat'
+        if expandpath(rc).exists():
+            self.args.extend(['/k', rc])
+        super(CmderCommandPrompt, self).fork(env, cwd)
+
+
+class CmderPowershell(MicrosoftShell):
+    def fork(self, env, cwd):
+        rc = '%CMDER_ROOT%\\vendor\\profile.ps1'
+        if expandpath(rc).exists():
+            self.args.extend([
+                '-ExecutionPolicy', 'Bypass', '-NoLogo', '-NoProfile',
+                '-NoExit', '-Command',
+                "Invoke-Expression '. ''{0}'''".format(rc),
+            ])
+        super(CmderPowershell, self).fork(env, cwd)
 
-def fork_cmder(env, cwd):
-    shell_cmd = ['cmd']
-    cmderrc_path = r'%CMDER_ROOT%\vendor\init.bat'
-    if expandpath(cmderrc_path).exists():
-        shell_cmd += ['/k', cmderrc_path]
-    if cwd:
-        os.environ['CMDER_START'] = cwd
-    fork_shell(env, shell_cmd, cwd)
 
 def _detect_shell():
```

Comment:
The decision tree in this function is kinda messy, wonder if we can clean this up any

Reply:
Not sure if this is better.

Reply:
Rewrote this a bit. Not sure if this makes more sense.